// eslint-plugin-hydrogen/dist/rules/no-state-in-server-components/tests/no-state-in-server-components.test.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const experimental_utils_1 = require("@typescript-eslint/experimental-utils");
const types_1 = require("@typescript-eslint/types");
const no_state_in_server_components_1 = require("../no-state-in-server-components");
const dedent_1 = __importDefault(require("dedent"));
const ruleTester = new experimental_utils_1.TSESLint.RuleTester({
    parser: require.resolve('@typescript-eslint/parser'),
});
function error(hookName) {
    return {
        type: types_1.AST_NODE_TYPES.CallExpression,
        data: { hook: hookName },
        messageId: 'noStateInServerComponents',
    };
}
ruleTester.run('hydrogen/no-state-in-server-components', no_state_in_server_components_1.noStateInServerComponents, {
    valid: [
        {
            code: dedent_1.default `
        function ClientComponent() {
          const [state, setState] = useState();
          return null;
        }
      `,
            filename: 'ClientComponent.client.tsx',
        },
        {
            code: dedent_1.default `
        function ClientComponent() {
          const [state, dispatch] = useReducer(() => {}, {});
          return null;
        }
      `,
            filename: 'ClientComponent.client.tsx',
        },
        {
            code: dedent_1.default `
        function ServerComponent() {
          return null;
        }
      `,
            filename: 'ServerComponent.server.tsx',
        },
        {
            code: dedent_1.default `
        function ServerComponent() {
          const {foo} = useBar();
          return null;
        }
      `,
            filename: 'ServerComponent.server.tsx',
        },
    ],
    invalid: [
        {
            code: dedent_1.default `
        function ServerComponent() {
          const [state, setState] = useState();
          return null;
        }
      `,
            errors: [error('useState')],
            filename: 'ServerComponent.server.tsx',
        },
        {
            code: dedent_1.default `
        function ServerComponent() {
          const [state, dispatch] = useReducer(() => {}, {});
          return null;
        }
      `,
            errors: [error('useReducer')],
            filename: 'ServerComponent.server.tsx',
        },
        {
            code: dedent_1.default `
        function ServerComponent() {
          const [state, dispatch] = useReducer(() => {}, {});
          const [state, setState] = useState();
          return null;
        }
      `,
            errors: [error('useReducer'), error('useState')],
            filename: 'ServerComponent.server.tsx',
        },
        {
            code: dedent_1.default `
        function ServerComponent() {
          const [state, setState] = useState();
          return null;
        }
      `,
            errors: [error('useState')],
            filename: 'ServerComponent.tsx',
        },
        {
            code: dedent_1.default `
        function ServerComponent() {
          const [state, dispatch] = useReducer(() => {}, {});
          return null;
        }
      `,
            errors: [error('useReducer')],
            filename: 'ServerComponent.tsx',
        },
        {
            code: dedent_1.default `
        function ServerComponent() {
          const [state, dispatch] = useReducer(() => {}, {});
          const [state, setState] = useState();
          return null;
        }
      `,
            errors: [error('useReducer'), error('useState')],
            filename: 'ServerComponent.tsx',
        },
    ],
});
